### PR TITLE
Better bash remediations tests.

### PR DESCRIPTION
**Pull request by matejak**
The ultimate goal of this PR is to test that bash remediations can be applied individually, and they will still succeed.
~Depends on https://github.com/OpenSCAP/scap-security-guide/pull/2468~

The PR:
- Implements checks for remediations done in separate steps.
- Adds checks for separate-step bash remediations.
- Reduces code duplication.

### CODE REVIEW

File: tests/README.md [hunk 1]
```diff
@@ -11,38 +11,45 @@ remediation works.
...
 If you want to use your own domain, make sure `openscap-1.2.15` and `qemu-quest-agent`
 are installed there, `root` is accessible via ssh and that `yum` command is able
 to install packages.
 
 ### Domain preparation (example CentOS)
-1. Install domain, using kickstarts/rhel_centos_7.cfg
-1. Import ssh key to the machine
-1. Setup repo, so machine can install and uninstall packages
+1. Install domain, using `kickstarts/rhel_centos_7.cfg`
+   Typically, you supply a boot option `inst.ks=https://raw.githubusercontent.com/matejak/scap-security-guide/master/tests/kickstarts/rhel_centos_7.cfg`.
```
Comment (dahaic) [inline tests/README.md]: Please refer to core project, not your fork :)

File: tests/test_suite.py [hunk 2]
```diff
@@ -55,11 +54,11 @@
...
-common_parser.add_argument("--ansible",
-                           dest="ansible",
-                           action="store_true",
-                           default=False,
-                           help="Use ansible instead of bash remediations")
+common_parser.add_argument("--remediate-using",
+                           dest="remediate_using",
+                           default="oscap",
+                           choices=ssg_test_suite.oscap.REMEDIATION_RULE_RUNNERS.keys(),
+                           help="What type of remediations to use.")
```
Comment (dahaic) [inline tests/test_suite.py]: It would help to explain what each choice means, at least what is a difference between `oscap` and `bash`

File: tests/analyze_test.py [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+import argparse
+
+NO_STATE_KEYS = ["name",]
+
+
+def get_table(records):
+    columns = tuple(sorted(get_columns(records)))
+    counts = []
+    for record in records:
+        counts.append((record.get('name', ''),) + tuple(get_values(record, columns)))
```
Comment (mpreisler) [inline tests/analyze_test.py]: needs some comments, this is pretty hard to read

File: tests/test-analyze_test.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
+import json
+import os
+
+import pytest
+
+import analyze_test as analyze
+
+
+def transform_table(table):
```
Comment (mpreisler) [inline tests/test-analyze_test.py]: seems like a waste to have this delegate defined on 3 lines

**Comment by mpreisler** [comment]
Didn't have time to do a full review but one common theme here is that the code has no comments even in areas where I would expect some. There is some code bloat where you define an analyze module that you only use once. Perhaps define it in the test itself and don't have multiple modules for it? If you want to have a module for it it needs docs and justification, is it going to be used anywhere else? Same with some of the delegates, python has lambdas so you don't have to define a function that just delegates to another function binding in an arg.

So tl;dr:
1) provide docs, when you manipulate tables either use an API that makes it obvious or show how you use lists of tuples of tuples as a table
2) make the code shorter, avoid delegates, avoid modules for things only used once
3) fix pep8 issues
4) provide a PR description so that it's easy for a reviewer to test this

**Comment by matejak** [comment]
I will make the code more understandable and comment it where appropriate. I don't get the comment about the `analyze` module - do you suggest to move it under the `ssg_test_suite` package?

To put this in a context, this PR is not a candidate for mergning. Although the code it is supposed to work, the essential pre-merge polish will be applied after @dahaic confirms that the functionality meets expectations. However, all comments are welcomed.

**Comment by mpreisler** [comment]
> To put this in a context, this PR is not a candidate for mergning

This absolutely has to be in the title then. It was not obvious to me that this is not for merging.

**Comment by mpreisler** [comment]
> about the analyze module - do you suggest to move it under the ssg_test_suite package?

Yeah, if it's a module that is only used at one place and the place where it's used is a small delegate executable I think we can just have it all there. Would be simpler and more readable. I could of course be missing something :)

**Comment by mpreisler** [comment]
> A separate script analyze_test has been introduced together with its tests test-analyze_test.py. It is supposed to display and compare results of earlier tests by reading those .json files. The ultimate goal is to make discovery of various irregularities or regressions easy and quickly provide the user with essential data.

I really wish that was a separate PR.

**Comment by dahaic** [comment]
sudo ./test_suite.py rule --hypervisor 'qemu:///system' --domain ssg-test-suite --datastream ../build/ssg-rhel7-ds.xml --benchmark xccdf_org.ssgproject.content_benchmark_RHEL-7 --remediate-using bash sshd

**Comment by matejak** [comment]
I have rebased the branch against recent master and took out the script for analysis.
The requested missing functionality was added, so if it passes, I will focus on formal aspects of the PR to prepare it for merging.

**Comment by matejak** [comment]
The PR is now considered a merge candidate. I have not fixed all line-length issues though. Although I agree that most lines should fit under 80 chars, I think that enforcing the limit for every line is counter-productive. However, if the general consensus is that we want to keep this restriction, I will fix the remaining lines.

Aside from implementing some new functionality, the existing code was heavily reorganized, so the diff doesn't make so much sense. I have updated the description with new highlights.

**Comment by dahaic** [comment]
Great job Matej!!! Test Suite is in great shape now.
